File: algorithm/leetcodeCpp/maximum-size-subarray-sum-equals-k.cpp
325. Maximum Size Subarray Sum Equals k
Difficulty : Easy 

Given an array nums and a target value k, find the maximum length of a subarray that sums to k. If there isnt one, return 0 instead.

Example 1:
Given nums = [1, -1, 5, -2, 3], k = 3,
return 4. (because the subarray [1, -1, 5, -2] sums to 3 and is the longest)

Example 2:
Given nums = [-2, -1, 2, 1], k = 1,
return 2. (because the subarray [-1, 2] sums to 1 and is the longest)

Follow Up:
Can you do it in O(n) time?

// Time:  O(n)
// Space: O(n)

class Solution {
public:
    int maxSubArrayLen(vector<int>& nums, int k) {
        unordered_map<int, int> sums;
        int sum = 0, maxLen = 0;
        for (int i = 0; i < nums.size(); ++i) {
            sum += nums[i];
            if (sum == k) {
                maxLen = i + 1;
            } else if (sums.count(sum - k)) {
                maxLen = max(maxLen, i - sums[sum - k]);
            }
            if (!sums.count(sum)) {
                sums[sum] = i;  // Only keep the smallest index.
            }
        }
        return maxLen;
    }
};

File: data-structure-and-algorithm/javascript/leetcode/maximum-size-subarray-sum-equals-k.js
/*
Given an array nums and a target value k, find the maximum length of a subarray that sums to k. If there isn't one, return 0 instead.

Note:
The sum of the entire nums array is guaranteed to fit within the 32-bit signed integer range.

Example 1:

Input: nums = [1, -1, 5, -2, 3], k = 3
Output: 4 
Explanation: The subarray [1, -1, 5, -2] sums to 3 and is the longest.
Example 2:

Input: nums = [-2, -1, 2, 1], k = 1
Output: 2 
Explanation: The subarray [-1, 2] sums to 1 and is the longest.
Follow Up:
Can you do it in O(n) time?
*/

function maxSubArrayLen(nums, k) {
    let sum = 0, result = 0;
    const map = new Map();
    for (let i = 0; i < nums.length; ++i) {
        sum += nums[i];
        if (sum === k) {
            result = i + 1;
        } else if (map.has(sum - k)) {
            result = Math.max(result, i - map.get(sum - k));
        }
        if (!map.has(sum)) {
            map.set(sum, i);
        }
    }
    return result;
}